added jamepad support

born2snipe opened the pull request.

born2snipe committed:
added jamepad support

MrStahlfelge commented:
Thank you! I will check this PR soon. Are you using this already in a game?

MrStahlfelge commented on desktop-jamepad/src/main/java/com/badlogic/gdx/controllers/desktop/DesktopControllerManager.java:
```diff
@@ -0,0 +1,10 @@
+import de.golfgl.gdx.controllers.jamepad.JamepadControllerManager;
+
+/**
+ * Total hack to override the default behavior of LibGDX to pickup this controller manager using "classpath magic".
```
I had planned to use the same hack to bring libGDX to use JamePad. 👍 
🥇

born2snipe commented:
No I am not using this in a game yet. I just came across your project here and I was unaware of Jamepad's existence. So I thought I would give the library a chance and see what the library could do. 

I really like the idea of the database for helping manage the mappings of the controllers. Reminds me of what [gamepad4j](https://github.com/marcelschoen/gamepad4j) was trying to do, except this lib directly depends on a well known native lib SDL2 vs. a hand rolled native lib.

Just a friendly FYI, the Jamepad's version of SDL2 (2.0.4) database file has changed over time. The controller mapping GUID's algorithms have [changed between versions](https://github.com/gabomdq/SDL_GameControllerDB#sdl-variants) of SDL2. So at some point, it may be beneficial to attempt to upgrade jamepad's SDL2 version. 

Also, if you do/don't like anything I have done or have questions, feedback/questions are welcome. 😁

MrStahlfelge commented on build.gradle:
```diff
@@ -8,6 +8,7 @@ buildscript {
     repositories {
         mavenCentral()
         maven { url "https://oss.sonatype.org/content/repositories/snapshots/" }
+        maven { url "http://maven.imagej.net/content/repositories/public/" }
```
For me, it does not work here.  The line must be added in the allprojects section around line 40.
What is the reason you did not use jitpack?

born2snipe replied:
The reason I used that [Maven repo](http://maven.imagej.net/content/repositories/public/) is because I generally use this [site](https://mvnrepository.com/artifact/com.github.WilliamAHartman/Jamepad/1.2) for looking up dependencies, and it said the dependency was listed in that repo.  

Sorry, I see now where you got the reference to [jitpack](https://github.com/williamahartman/Jamepad/#getting-jamepad).

MrStahlfelge commented on desktop-jamepad/src/main/java/de/golfgl/gdx/controllers/jamepad/support/JamePadExceptions.java:
```diff
@@ -0,0 +1,7 @@
+package de.golfgl.gdx.controllers.jamepad.support;
```
this file is unneccessary (see comments later)

MrStahlfelge commented on desktop-jamepad/src/main/java/de/golfgl/gdx/controllers/jamepad/support/JamepadController.java:
```diff
@@ -0,0 +1,196 @@
+
+    @Override
+    public PovDirection getPov(int povCode) {
+        throw notSupported("getPov(...)");
```
This exception must not be thrown because it does thwarts the whole point about being platform-independant. A game calling this method will crash. The method is called by controller-utils-mapping as well. ;-)

The GWT implementation returned 
    return PovDirection.center;
before POV support was added. That's the right thing to do here.

MrStahlfelge commented on desktop-jamepad/src/main/java/de/golfgl/gdx/controllers/jamepad/support/JamepadController.java:
```diff
@@ -0,0 +1,196 @@
+    @Override
+    public boolean getSliderX(int sliderCode) {
+        throw notSupported("getSliderX(...)");
+    }
```
return false for same reason

MrStahlfelge commented on desktop-jamepad/src/main/java/de/golfgl/gdx/controllers/jamepad/support/JamepadController.java:
```diff
@@ -0,0 +1,196 @@
+    @Override
+    public boolean getSliderY(int sliderCode) {
+        throw notSupported("getSliderY(...)");
+    }
```
return false (see above)

MrStahlfelge commented on desktop-jamepad/src/main/java/de/golfgl/gdx/controllers/jamepad/support/JamepadController.java:
```diff
@@ -0,0 +1,196 @@
+    @Override
+    public Vector3 getAccelerometer(int accelerometerCode) {
+        throw notSupported("getAccelerometer(...)");
+    }
```
return Vector3.Zero;

MrStahlfelge commented on desktop-jamepad/src/main/java/de/golfgl/gdx/controllers/jamepad/support/JamepadController.java:
```diff
@@ -0,0 +1,196 @@
+    @Override
+    public void setAccelerometerSensitivity(float sensitivity) {
+        throw notSupported("setAccelerometerSensitivity(...)");
+    }
```
do nothing

MrStahlfelge commented on desktop-jamepad/src/main/java/de/golfgl/gdx/controllers/jamepad/support/JamepadController.java:
```diff
@@ -0,0 +1,196 @@
+        return query(new ControllerQuerier<Boolean>() {
+            @Override
+            public Boolean query(ControllerIndex controllerIndex) throws ControllerUnpluggedException {
+                return controllerIndex.isButtonPressed(toButton(buttonCode));
```
This can lead to a NPE if an unknown buttonCode is given.

MrStahlfelge commented on desktop-jamepad/src/main/java/de/golfgl/gdx/controllers/jamepad/support/JamepadController.java:
```diff
@@ -0,0 +1,196 @@
+    public float getAxis(final int axisCode) {
+        return query(new ControllerQuerier<Float>() {
+            public Float query(ControllerIndex controllerIndex) throws ControllerUnpluggedException {
+                return controllerIndex.getAxisState(toAxis(axisCode));
```
This can lead to a NPE if an unknown axisCode is given

MrStahlfelge committed:
Jamepad Integration: some fixes

MrStahlfelge reviewed (changes requested):
First, thank you very much for your work. I really appreciate it and I was able to control my desktop build with this. 

I've checked the lib and made some comments. I have pulled your commit to the /jamepad branch in my repo and fixed the commented changes.

There is a crash if you unplug a controller while the game is running (nested iterations). I did not track the reason for this down. Also, of the 3 controllers I connected only one was recognized. I don't know if this is a problem of the gamepad db.

born2snipe committed:
give reasonable defaults for unsupported functionality of jamepad

born2snipe committed:
handle potential invalid button/axis codes

born2snipe committed:
changed to use the proper maven repo for downloading jamepad

born2snipe commented:
@MrStahlfelge I believe I have addressed all the issues.

MrStahlfelge commented:
Hi,

thanks for the changes.

For me the reference to the Maven repository does not work where it is. It needs to be placed inside allprojects { ...} 

What about the following problems:

> There is a crash if you unplug a controller while the game is running (nested iterations). I did not track the reason for this down. Also, of the 3 controllers I connected only one was recognized. I don't know if this is a problem of the gamepad db.

born2snipe commented:
The changes for the repo stuff should be pushed now.

It would be interesting to know if you use [this](http://www.generalarcade.com/gamepadtool/) tool and see if the tool recognizes your controllers. The unfortunate thing is the tool is on a newer version of SDL2.

What OS are you on when you have the crashes?

MrStahlfelge commented on desktop-jamepad/src/main/java/de/golfgl/gdx/controllers/jamepad/support/JamepadController.java:
```diff
@@ -0,0 +1,196 @@
+            try {
+                return querier.query(controllerIndex);
+            } catch (ControllerUnpluggedException e) {
+                logger.error("Failed querying controller at index: " + controllerIndex.getIndex(), e);
```
This line causes a lot of redundant information written to the error log when a controller is disconnected. Should better be just an logger.info without all the stacktrace stuff.

born2snipe replied:
The excessive logging smells like a different problem, meaning, it sounds like when the controller is disconnected we are not properly disposing of the controller and probably should not be querying the controller if it has been disconnected. Does that make sense?

MrStahlfelge replied:
I think the querying is just done before the controller disconnect callback is done. The game crashes on the disconnect callback which comes half a second after the controller is disconnected. Why do you call reconnectController()?

born2snipe replied:
The hope was potentially reconnecting the controller and being able to continue. If you think that is overkill, I am cool with that going away.

MrStahlfelge commented:
I've added another comment on JamepadController, and here is the stacktrace that causes the game to crash when a controller is disconnected. I am on Windows.

`
Exception in thread "LWJGL Application" com.badlogic.gdx.utils.GdxRuntimeException: #iterator() cannot be used nested.
	at com.badlogic.gdx.utils.Array$ArrayIterator.hasNext(Array.java:577)
	at de.golfgl.gdx.controllers.jamepad.support.CompositeControllerListener.notifyListeners(CompositeControllerListener.java:90)
	at de.golfgl.gdx.controllers.jamepad.support.CompositeControllerListener.axisMoved(CompositeControllerListener.java:46)
	at de.golfgl.gdx.controllers.jamepad.support.CompositeControllerListener$3.perform(CompositeControllerListener.java:48)
	at de.golfgl.gdx.controllers.jamepad.support.CompositeControllerListener.notifyListeners(CompositeControllerListener.java:91)
	at de.golfgl.gdx.controllers.jamepad.support.CompositeControllerListener.axisMoved(CompositeControllerListener.java:46)
	at de.golfgl.gdx.controllers.jamepad.support.JamepadController.updateAxisState(JamepadController.java:144)
	at de.golfgl.gdx.controllers.jamepad.support.JamepadController.update(JamepadController.java:124)
	at de.golfgl.gdx.controllers.jamepad.support.JamepadControllerMonitor.run(JamepadControllerMonitor.java:37)
	at com.badlogic.gdx.backends.lwjgl.LwjglApplication.executeRunnables(LwjglApplication.java:259)
	at com.badlogic.gdx.backends.lwjgl.LwjglApplication.mainLoop(LwjglApplication.java:210)
	at com.badlogic.gdx.backends.lwjgl.LwjglApplication$1.run(LwjglApplication.java:126)
`

About the controllers not getting recognized, I will check later. I use some SDL games that are able to use the controllers, but that might be also because they use a newer version.

born2snipe committed:
added a simple test ui to see the controller inputs working or not

born2snipe committed:
remove logging noise and retrying

born2snipe commented:
Unfortunately I am away from my home and I do not have access to a Windows machine. I am currently on a Mac. 

I added a simple GUI for testing the inputs from the controllers. I tested with a "wired xbox 360" and a "iBuffalo Controller" and at least on Mac, I am failing to re-create the error your described earlier. Maybe there is a race condition I am not seeing.

MrStahlfelge committed:
Merge remote-tracking branch 'born2snipe/master' into jamepad

# Conflicts:
#	desktop-jamepad/src/main/java/de/golfgl/gdx/controllers/jamepad/support/JamepadController.java

MrStahlfelge committed:
controllers-jamepad: Readme, move controllerdb to main folder

MrStahlfelge commented:
The crash is gone now. 👍 
Unfortunately, I discovered a new problem I could not see because of the crash: If I disconnect a controller, any new connected controller is not reported until I restart the session. This problem just occurs if a game controller was disconnected before.

The gamepadtool recognizes all of my controllers, I am also testing with the iBuffalo SNES controller and it is one of the controllers not working. 😄 I tracked down the problem and could solve it. The gamecontrollerdb.txt file must not sit under a jamepad/ subdir, but must be added to the main resources directory. This is due to a bug in Jamepad when copying the file to the temp directory which does not work (at least on Windows) if the target destination folder (which was temp/jamepad/) does not exist. I pushed a fix to my /jamepad branch

It is interesting that the controllers are not reported at all if the mapping is not found.

By the way, with Android Studio/IntelliJ I am not able to start the test game. It must be located in an own subproject, but I prefer using an own example project.

born2snipe commented:
I just pulled in your 2 commits and re-pushed.

> It is interesting that the controllers are not reported at all if the mapping is not found.

I agree. That seems problematic if someone wants to use a new controller that is unrecognized and it could be simple enough to add a "controller mapping" screen so a game/app would support their controller.

> By the way, with Android Studio/IntelliJ I am not able to start the test game. It must be located in an own subproject, but I prefer using an own example project.

You want the test tool in it's own module/project?

born2snipe committed:
moved the jamepad test app to it's own module

born2snipe committed:
fixed problem of the controller monitoring thread not properly clearing state when a controller was disconnected

Also moved the calling of the listener for disconnections to the same place (ie. JamepadControllerMonitor) as the connection notifications. This was done to be consistent so the connection management is done by the thread and all the input notifications are done by the controller itself.

born2snipe committed:
fix threading issue in jamepad test app of plugging in/out of multiple controllers

born2snipe committed:
fix issue in the test app not properly resetting the controller name list on disconnects

MrStahlfelge committed:
Jamepad: Reflip vertical axis values to be consistent with other platforms

born2snipe commented:
Based on my testing with the test app on Mac, I am experiencing good plug n' play behavior. Please let me know if you are having better luck.

MrStahlfelge committed:
Updating readme for Jamepad

MrStahlfelge commented:
Thank you. Everything's working well now!

There's just one question open:
Looking at your last commit to ensure threading issues: libGDX own controller implementations ensure the ControllerAdapters are only called on the main thread. This is mandatory, and I thought it is done here also by calling JamepadControllerMonitor's run() method on the main thread. So either I don't understand correctly, or the synchronizing statements are not necessary and will confuse people checking the test app's code.

I've also made an important commit on my branch. Jamepad flips vertical axis values. It is commented in the code: 
     `//Note: we flip the Y values so up on the stick is positive. that makes more sense.`
It's great that the author thinks it makes more sense this way. 😆 Unfortunately Android, WebGL and even SDL native does it the other way round and to remain platform independant, I've added code to reflip the values.

Maybe you want to join the libGDX discord channel: https://discord.gg/2pdra8D

born2snipe commented:
ok so I found out what my problem is...I have no idea if this affects windows or if it is isolated to mac only.

Steps that usually re-create the problem: 
1. plug 1st controller in USB-1
2. plug 2nd controller in USB-2
3. unplug 1st controller

What I am seeing happen is the jamepad is reassigning the 2nd controller to the 1st controller's instance and index. So from the test app you will see the 1st controller's name in the list, but you see the inputs from the 2nd controller. I am not exactly sure how to fix this on our end. This seems like a bug in jamepad or there needs to be a real unique id for a controller coming from jamepad.

If you got other ideas, please let me know. 😄 

FYI: If you would like to see what I am seeing or at least try to see what I am seeing here are the changes to make: 

All inside the `JamepadControllerMonitor.java`:
```java
private void checkForNewControllers() {
        System.out.println("################## checkForNewControllers");
        int newNumControllers = controllerManager.getNumControllers();
        for (int i = 0; i < newNumControllers; i++) {
            ControllerIndex controllerIndex = controllerManager.getControllerIndex(i);

            try {
                System.out.println(controllerIndex.getIndex() + " -- "+controllerIndex.getName()+" -- "+controllerIndex);
            } catch (ControllerUnpluggedException e) {

            }

            if (!indexToController.containsKey(controllerIndex.getIndex())) {
                Tuple tuple = new Tuple(controllerIndex);
                tuple.controller.addListener(listener);

                indexToController.put(controllerIndex.getIndex(), tuple);
                listener.connected(tuple.controller);
            }
        }
        System.out.println("######################################################");
    }
```

You should see something like below:
```
################## checkForNewControllers
0 -- iBuffalo USB 2-axis 8-button Gamepad -- com.studiohartman.jamepad.ControllerIndex@690aa03a
1 -- X360 Controller -- com.studiohartman.jamepad.ControllerIndex@282f40d6
######################################################
################## checkForNewControllers
0 -- X360 Controller -- com.studiohartman.jamepad.ControllerIndex@690aa03a
######################################################
```

born2snipe commented:
Also, you are correct in the threading, the `synchronized` are *not* needed.  I just made the wrong assumption of seeing the controller names changing in the GUI being a threading issue.

born2snipe commented:
I didn't realize `jamepad` came with it's own tester (ie. `ControllerTester.java`) and it is showing the same behavior.

MrStahlfelge commented:
The behaviour is correct in my opinion, and it is the way it works on other platforms, too. I did not discover this problem because my game deals with this by clearing and refreshing the list. 
The libGDX select box used in the test app also has some optimizations that interfere with this, in my experience you have to be very careful when using Scene2d lists.
I changed the test app to refresh the select box from scratch on every connect and disconnect and this works well. The focus is lost and you have to reselect the controller, but I think that is really a minor problem.

MrStahlfelge commented:
I've pulled everything besides the latest two changes on the test app that are in conflict with my fix and close this PR. Thank you very much for all your work, it is very appreciated (and I am sorry that Github shows this PR as rejected now, which is definitely not correct). 

We can deal with further problems with the Jamepad implementation in dedicated issues. Let me know when you think I can push a release version of this.

I have made a Wiki page with two items "work to be done".

MrStahlfelge commented:
> Just a friendly FYI, the Jamepad's version of SDL2 (2.0.4) database file has changed over time. The controller mapping GUID's algorithms have [changed between versions](https://github.com/gabomdq/SDL_GameControllerDB#sdl-variants) of SDL2. So at some point, it may be beneficial to attempt to upgrade jamepad's SDL2 version.

Jamepad was updated now with v1.3
